### PR TITLE
refactor(`zetaclient`): improve some general structure of the codebase

### PR DESCRIPTION
**Pull request by lumtis**
# Description

Add some improvement in the codebase following this issue: https://github.com/zeta-chain/node/issues/2026

### CODE REVIEW

File: zetaclient/bitcoin/bitcoin_client.go [hunk 1]
```diff
@@ -61,33 +71,54 @@ type BTCLog struct {
...
-	BlockTime        uint64 // block time in seconds
+	pendingNonce     uint64
+	utxos            []btcjson.ListUnspentResult
+	params           observertypes.ChainParams
+	coreContext      *corecontext.ZetaCoreContext
+
+	// includedTxHashes indexes included tx with tx hash
+	includedTxHashes map[string]bool
+
+	// includedTxResults indexes tx resullts with the oubound nonce
```
Comment (ws4charlie) [inline zetaclient/bitcoin/bitcoin_client.go:112]: // includedTxResults indexes tx resullts with the oubound tx identifier

File: zetaclient/bitcoin/bitcoin_client.go [hunk 2]
```diff
@@ -61,33 +71,54 @@ type BTCLog struct {
...
 
-	Mu                *sync.Mutex // lock for all the maps, utxos and core params
-	pendingNonce      uint64
-	includedTxHashes  map[string]bool                          // key: tx hash
-	includedTxResults map[string]*btcjson.GetTransactionResult // key: chain-tss-nonce
-	broadcastedTx     map[string]string                        // key: chain-tss-nonce, value: outTx hash
-	utxos             []btcjson.ListUnspentResult
-	params            observertypes.ChainParams
-	coreContext       *corecontext.ZetaCoreContext
+	// broadcastedTx indexes the outbound hash with the TSS nonce
```
Comment (ws4charlie) [inline zetaclient/bitcoin/bitcoin_client.go:115]: // broadcastedTx indexes the outbound hash with the outbound tx identifier

File: zetaclient/bitcoin/bitcoin_client.go [hunk 3]
```diff
@@ -170,12 +206,15 @@ func NewBitcoinClient(
...
+
+	// set the Bitcoin chain params
 	_, chainParams, found := appcontext.ZetaCoreContext().GetBTCChainParams()
 	if !found {
 		return nil, fmt.Errorf("btc chains params not initialized")
 	}
 	ob.params = *chainParams
-	// initialize the Client
+
+	// connect to the Client
```
Comment (ws4charlie) [inline zetaclient/bitcoin/bitcoin_client.go:217]:
```suggestion
	// create the RPC client
```

File: zetaclient/evm/evm_client.go [hunk 4]
```diff
@@ -317,8 +316,8 @@ func (ob *ChainClient) Stop() {
 	ob.logger.Chain.Info().Msgf("%s observer stopped", ob.chain.String())
 }
 
-// returns: isIncluded, isConfirmed, Error
-// If isConfirmed, it also post to ZetaCore
+// IsSendOutTxProcessed returns isIncluded, isConfirmed, Error
+// ff isConfirmed, it also post to ZetaCore
```
Comment (ws4charlie) [inline zetaclient/evm/evm_client.go:320]:
```suggestion
// if isConfirmed, it also post to ZetaCore
```

File: zetaclient/testutils/testdata.go [hunk 5]
```diff
@@ -24,19 +24,6 @@ const (
 	RestrictedBtcAddressTest = "bcrt1qzp4gt6fc7zkds09kfzaf9ln9c5rvrzxmy6qmpp"
 )
 
-// SaveObjectToJSONFile saves an object to a file in JSON format
```
Comment (ws4charlie) [inline zetaclient/testutils/testdata.go:27]: Can we keep `SaveObjectToJSONFile`? It's been used to save downloaded JSON-based test data.

File: zetaclient/testutils/testdata.go [hunk 6]
```diff
@@ -55,27 +42,6 @@ func ComplianceConfigTest() config.ComplianceConfig {
 	}
 }
 
-// SaveTrimedEVMBlockTrimTxInput trims tx input data from a block and saves it to a file
-func SaveEVMBlockTrimTxInput(block *ethrpc.Block, filename string) error {
```
Comment (ws4charlie) [inline zetaclient/testutils/testdata.go:59]: Same. It's been used to save trimmed block.

File: zetaclient/testutils/testdata.go [hunk 7]
```diff
@@ -55,27 +42,6 @@ func ComplianceConfigTest() config.ComplianceConfig {
...
-	for i := range block.Transactions {
-		block.Transactions[i].Input = "0x"
-	}
-	return SaveObjectToJSONFile(block, filename)
-}
-
-// SaveTrimedBTCBlockTrimTx trims tx data from a block and saves it to a file
-func SaveBTCBlockTrimTx(blockVb *btcjson.GetBlockVerboseTxResult, filename string) error {
```
Comment (ws4charlie) [inline zetaclient/testutils/testdata.go:67]: same

**Review by ws4charlie — COMMENTED** [review]
looks nice. Left a few minor comments.

File: zetaclient/testutils/testdata.go (hunk 5 above)
Comment (lumtis) [reply zetaclient/testutils/testdata.go:27]: Ok, I will add back and add comment for context

Comment (lumtis) [reply zetaclient/testutils/testdata.go:27]: 6e57c8f9

File: zetaclient/testutils/testdata.go (hunk 6 above)
Comment (lumtis) [reply zetaclient/testutils/testdata.go:59]: 6e57c8f9

File: zetaclient/bitcoin/bitcoin_client.go [hunk 8]
```diff
@@ -61,33 +71,54 @@ type BTCLog struct {
 	Compliance zerolog.Logger // The logger for compliance checks
 }
 
+// BTCInTxEvent represents an incoming transaction event
+type BTCInTxEvent struct {
+	// FromAddress is the first input address
+	FromAddress string
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/bitcoin/bitcoin_client.go:77]: I have no preference for one over the other (inline or above line). I think code is generally read top-to-bottom, maybe we can standardize the above-line comments?

File: zetaclient/bitcoin/bitcoin_client.go [hunk 9]
```diff
@@ -413,80 +428,79 @@ func (ob *BTCChainClient) ObserveInTx() error {
...
+	blockNumber := lastScanned + 1
+	res, err := ob.GetBlockByNumberCached(blockNumber)
+	if err != nil {
+		ob.logger.InTx.Error().Err(err).Msgf("observeInTxBTC: error getting bitcoin block %d", blockNumber)
+		return err
+	}
+	ob.logger.InTx.Info().Msgf("observeInTxBTC: block %d has %d txs, current block %d, last block %d",
+		blockNumber, len(res.Block.Tx), cnt, lastScanned)
+
+	// print some debug information
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/bitcoin/bitcoin_client.go:440]: If you like, I think we can simply remove the log print. My guess is that the log print here was only used in the early days for debugging when we initially played and integrated Bitcoin. Also, it prints thousands of txs in the block which is too much.

**Review by ws4charlie — APPROVED** [review]

File: zetaclient/bitcoin/bitcoin_client.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (lumtis) [reply zetaclient/bitcoin/bitcoin_client.go:88]: Changed for the ZetaClient 0120722d

File: zetaclient/bitcoin/bitcoin_client.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (lumtis) [reply zetaclient/bitcoin/bitcoin_client.go:456]: Let's remove

File: zetaclient/config/types_test.go [hunk 10]
```diff
@@ -1 +0,0 @@
-package config_test
```
[the remark this answers is not in the document]
Comment (lumtis) [reply zetaclient/config/types_test.go:1]: Good point

File: zetaclient/evm/evm_client.go [hunk 11]
```diff
@@ -31,30 +34,21 @@ import (
...
-}
-
-type OutTx struct {
-	SendHash string
-	TxHash   string
-	Nonce    int64
-}
-type Log struct {
+// Logger is the logger for evm chains
+type Logger struct {
```
[the remark this answers is not in the document]
Comment (lumtis) [reply zetaclient/evm/evm_client.go:51]: We can do this here: https://github.com/zeta-chain/node/issues/2022

I will add comment in the code to point the issue

File: zetaclient/evm/evm_signer.go [hunk 12]
```diff
@@ -488,6 +495,115 @@ func (signer *Signer) BroadcastOutTx(
...
+
+func (signer *Signer) EvmClient() interfaces.EVMRPCClient {
+	return signer.client
+}
+
+func (signer *Signer) EvmSigner() ethtypes.Signer {
+	return signer.ethSigner
+}
+
+// ________________________
```
[the remark this answers is not in the document]
Comment (lumtis) [reply zetaclient/evm/evm_signer.go:574]: No

File: zetaclient/evm/evm_signer.go [hunk 13]
```diff
@@ -488,6 +495,115 @@ func (signer *Signer) BroadcastOutTx(
...
+		return nil, fmt.Errorf("Sign error: %w", err)
+	}
+
+	return tx, nil
+}
+
+// Exported for unit tests
+
+func (signer *Signer) GetReportedTxList() *map[string]bool {
+	return &signer.outTxHashBeingReported
```
[the remark this answers is not in the document]
Comment (lumtis) [reply zetaclient/evm/evm_signer.go:563]: Yep, creating a specific issue for this: https://github.com/zeta-chain/node/issues/2084

File: zetaclient/evm/evm_signer.go [hunk 14]
```diff
@@ -488,6 +495,115 @@ func (signer *Signer) BroadcastOutTx(
...
+	tx, _, _, err := signer.Sign(data, txData.to, txData.gasLimit, txData.gasPrice, outboundParams.OutboundTxTssNonce, txData.height)
+	if err != nil {
+		return nil, fmt.Errorf("sign error: %w", err)
+	}
+	return tx, nil
+}
+
+func (signer *Signer) SignMigrateTssFundsCmd(txData *OutBoundTransactionData) (*ethtypes.Transaction, error) {
+	outboundParams := txData.outboundParams
+	tx, _, _, err := signer.Sign(nil, txData.to, txData.gasLimit, txData.gasPrice, outboundParams.OutboundTxTssNonce, txData.height)
```
[the remark this answers is not in the document]
Comment (lumtis) [reply zetaclient/evm/evm_signer.go:607]: This is a simple transfer

File: zetaclient/interfaces/signer.go [hunk 15]
```diff
@@ -16,8 +16,11 @@ import (
 
 type TSSSigner interface {
 	Pubkey() []byte
-	// Sign: Specify optionalPubkey to use a different pubkey than the current pubkey set during keygen
+
+	// Sign signs the data
+	// Note: it specifies optionalPubkey to use a different pubkey than the current pubkey set during keygen
```
[the remark this answers is not in the document]
Comment (lumtis) [reply zetaclient/interfaces/signer.go:21]: https://github.com/zeta-chain/node/issues/2085

File: zetaclient/evm/evm_client.go [hunk 16]
```diff
@@ -325,6 +325,300 @@ func (ob *ChainClient) Stop() {
...
+func (ob *ChainClient) IsSendOutTxProcessed(cctx *crosschaintypes.CrossChainTx, logger zerolog.Logger) (bool, bool, error) {
+	sendHash := cctx.Index
+	cointype := cctx.InboundTxParams.CoinType
+	nonce := cctx.GetCurrentOutTxParam().OutboundTxTssNonce
+
+	// skip if outtx is not confirmed
+	params := ob.GetChainParams()
+	receipt, transaction := ob.GetTxNReceipt(nonce)
+	if receipt == nil || transaction == nil { // not confirmed yet
+		return false, false, nil
```
[the remark this answers is not in the document]
Comment (lumtis) [reply zetaclient/evm/evm_client.go:360]: Let's iterate all the necessary logs once we start the monitoring epic

File: zetaclient/evm/evm_client.go [hunk 17]
```diff
@@ -325,6 +325,300 @@ func (ob *ChainClient) Stop() {
...
+	}
+
+	sendID := fmt.Sprintf("%s-%d", ob.chain.String(), nonce)
+	logger = logger.With().Str("sendID", sendID).Logger()
+
+	// compliance check, special handling the cancelled cctx
```
[the remark this answers is not in the document]
Comment (lumtis) [reply zetaclient/evm/evm_client.go:366]: I would avoid touching too deep the workflow logic in this PR, we'll check with Charlie on this

**Pull request merged**